src/seq_analysis/orf.rs

```diff
@@ -0,0 +1,152 @@
+//! let finder = NaiveFinder::new(start_codons, stop_codons, min_len);
+//!
+//! let sequence = b"ACGGCTAGAAAAGGCTAGAAAA"
+//!
+//! for orf in finder.find_all(sequence) {
+//!    ...do something...
+//! }
+//! ```
+
+//! Right now the only way to check the reverse strand for orf is to use
+//! the alphabet::dna::RevComp struct and to check for both sequences.
+//! But that's not so performance friendly, as the reverse complementation and the orf research
+//! could go on at the same time.
+
+
+/// Finder algorithm
+pub struct NaiveFinder <'a> {
+    start_codons: Vec<&'a [u8; 3]>,
+    stop_codons: Vec<&'a [u8; 3]>,
+    min_len: usize,
+}
+
+impl<'a> NaiveFinder <'a> {
+    /// Create a new instance for given genetic datas
+    pub fn new (start_codons: Vec<&'a [u8; 3]>, stop_codons: Vec<&'a [u8; 3]>, min_len: usize) -> Self {
+        NaiveFinder {
+            start_codons: start_codons,
+            stop_codons: stop_codons,
+            min_len: min_len,
+        }
+    }
+    /// Find all open reading frames in given sequence. Matches are returned as iterators over vectors
+    pub fn find_all(&'a self, sequence: &'a [u8]) -> Matches {
+        Matches::new(self, sequence)
+    }
+}
+
+/// container struct for argument iteration
+struct State {
+    index: usize,
+    length: usize,
+    start_pos: [usize; 3],
+    end_pos: [usize; 3],
+    in_orf: [bool; 3]
+}
+
+impl State {
+    /// Create new state
+    pub fn new(length: usize) -> Self {
+        State {
+            index: 0,
+            length: length,
+            start_pos: [0, 0, 0],
+            end_pos: [0, 0, 0],
+            in_orf: [false, false, false],
+        }
+    }
+}
+
+/// Iterator yielding open reading frames as vector (start_pos, end_pos + 1, sequence)
+pub struct Matches<'a> {
+    finder: &'a NaiveFinder<'a>,
+    sequence: &'a [u8],
+    state: State,
+}
+
+impl<'a> Matches<'a> {
+    pub fn new (finder: &'a NaiveFinder, sequence: &'a [u8]) -> Self {
+
+        //let mut v = Vec::new();
+        //v.extend_from_slice(sequence);
+
+        Matches {
+            finder: finder,
+            sequence: sequence,
+            state: State::new(sequence.len()),
+        }
+    }
+}
+
+
+impl<'a> Iterator for Matches<'a> {
+    type Item = (usize, usize, &'a [u8]);
+
+    fn next(&mut self) -> Option<(usize, usize, &'a [u8])> {
+
+        let mut orf: Option<(usize, usize, &'a [u8])> = None;
+        let mut codons: [ [u8; 3]; 3]; //
+
+        while self.state.index < self.state.length - 3 {
+            //codons array depending on reading frame
+            codons = [
+                [ self.sequence[self.state.index]  , self.sequence[self.state.index+1], self.sequence[self.state.index+2] ],
+                [ self.sequence[self.state.index+1], self.sequence[self.state.index+2], self.sequence[self.state.index+3] ],
+                [ self.sequence[self.state.index+2], self.sequence[self.state.index+3], self.sequence[self.state.index+4] ],
+            ];
```

Comment:
Working on it, thanks for the hints :)
Le 29 mars 2016 20:34, "Johannes Köster" <notifications@github.com> a
écrit :

> In src/seq_analysis/orf.rs
> <https://github.com/rust-bio/rust-bio/pull/48#discussion_r57775970>:
>
> > +
> > +impl<'a> Iterator for Matches<'a> {
> > +    type Item = (usize, usize, &'a [u8]);
> > +
> > +    fn next(&mut self) -> Option<(usize, usize, &'a [u8])> {
> > +
> > +        let mut orf: Option<(usize, usize, &'a [u8])> = None;
> > +        let mut codons: [ [u8; 3]; 3]; //
> > +
> > +        while self.state.index < self.state.length - 3 {
> > +            //codons array depending on reading frame
> > +            codons = [
> > +                [ self.sequence[self.state.index]  , self.sequence[self.state.index+1], self.sequence[self.state.index+2] ],
> > +                [ self.sequence[self.state.index+1], self.sequence[self.state.index+2], self.sequence[self.state.index+3] ],
> > +                [ self.sequence[self.state.index+2], self.sequence[self.state.index+3], self.sequence[self.state.index+4] ],
> > +            ];
>
> This is even better: You maintain a deque
> <https://doc.rust-lang.org/collections/vec_deque/struct.VecDeque.html> of
> the last 5 nucleotides in the struct. In each step, you
> self.codons.pop_front() the first nucleotide, and self.codons.push_back()
> the new nucleotide. It should also allow you to remove the dependence on
> having a slice of text, and allow you to use a IntoTextIterator instead.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-bio/rust-bio/pull/48/files/5ccc54036a28e80c7729883d00d2431991da01c7#r57775970>
>